### PR TITLE
Feature/user group ui

### PR DESCRIPTION
**Pull request by esotrope**
Not satisfied with the UX but it gets the job done. However, I figured this would be a good starting point for discussion.

TODOs: integrate advanced filters component as appears on the exam result page

image

image

image

### CODE REVIEW

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java [hunk 1]
```diff
@@ -100,7 +100,7 @@ public Integer getStudentCount() {
     /**
      * @return <code>true</code> if the group was created by a user instead of an admin for the user
      */
-    private boolean isUserCreated() {
+    public boolean isUserCreated() {
```
Comment (esotrope) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java]: bugfix

File: webapp/src/main/webapp/src/app/student/search/student-search-form.component.html [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+          <option [ngValue]="undefined"></option>
+          <option *ngFor="let group of options.groups"
+                  [ngValue]="group">{{group.name}}</option>
+        </select>
+      </div>
+    </div>
+
+    <div class="flex-child grow">
+      <div class="form-group">
+        <label>Student Name / SSID</label>
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/student/search/student-search-form.component.html]: plan on cleaning up the hardcoded text once the layout etc is more final

File: webapp/src/main/webapp/src/app/user-group/user-group.component.ts [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+      .replace(/[^\w\s]/g, '')
+      .replace(/\s+/g, '');
+
+    this.filteredStudents = this.students
+      .filter(student => {
+        return this.group.students.find(x => x.id === student.id) == null
+          && (
+            student.ssid.startsWith(this.studentForm.name)
+            || `${(student.lastName + student.firstName).toLowerCase()}`.includes(nameSearch)
+            || `${(student.firstName + student.lastName).toLowerCase()}`.includes(nameSearch)
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/user-group/user-group.component.ts]: implemented the student search by name/ssid on the client side with a somewhat forgiving matching strategy, not sure what we think about this.

File: webapp/src/main/webapp/src/app/user-group/user-group.component.ts [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+              private translate: TranslateService,
+              private notificationService: NotificationService) {
+    this.originalGroup = this.route.snapshot.data[ 'group' ];
+  }
+
+  get saveButtonDisabled(): boolean {
+    return this.processingSubscription != null
+      || this.groupForm == null
+      || !this.groupForm.formGroup.valid
+      || equals(this.originalGroup, this.group);
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/user-group/user-group.component.ts]: only lets you save if there was a change to the group

File: webapp/src/main/webapp/src/app/user-group/user-group.component.ts [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+  }
+
+  onSaveButtonClick(): void {
+    if (this.saveButtonDisabled) {
+      return;
+    }
+    // TODO only save when there are changes
+    this.processingSubscription = this.service.saveGroup(this.group)
+      .subscribe(() => {
+          this.navigateHome();
```
Comment (esotrope) [inline webapp/src/main/webapp/src/app/user-group/user-group.component.ts]: navigates the user away from the form on save delete and cancel to combat the "apply every change" kind of use

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java [hunk 6]
```diff
@@ -163,8 +163,8 @@ public Group build() {
             group.school = school;
             group.subjectCode = subjectCode;
             group.subjectCodes = subjectCodes != null && !subjectCodes.isEmpty() ? ImmutableSet.copyOf(subjectCodes) : null;
-            group.studentIds = studentIds != null ? ImmutableSet.copyOf(studentIds) : ImmutableSet.of();
-            group.students = students != null ? ImmutableSet.copyOf(students) : ImmutableSet.of();
+            group.studentIds = studentIds != null ? ImmutableSet.copyOf(studentIds) : null;
+            group.students = students != null ? ImmutableSet.copyOf(students) : null;
```
Comment (agorina) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/Group.java]: Maybe add a helper to `org.opentestsystem.rdw.reporting.common.util.CollectionUtils`?

**Comment by esotrope** [comment]
* Redesigned layout
* Added advanced filters
* Fixed form validation

Still todo:
1. Add invalid field text
2. Add delete modal

image

image

image

image

**Pull request merged**